contracts/command-batcher/src/encoding/evm.rs

```diff
@@ -0,0 +1,557 @@
+                let pub_key = pub_keys
+                    .get(&participant.address.to_string())
+                    .expect("violated invariant: participant address not found in pub_keys")
+                    .into();
+
+                Operator {
+                    address: evm_address(pub_key),
+                    weight: participant.weight.into(),
+                    signature: signers.get(participant.address.as_str()).cloned(),
+                }
+            })
+            .collect::<Vec<Operator>>();
+        operators.sort_by(|a, b| a.address.cmp(&b.address));
+
+        Proof {
+            operators,
+            threshold: snapshot.quorum.into(),
+        }
+    }
+
+    fn encode_execute_data(&self, data: &Data) -> HexBinary {
+        let data_encoded = data.encode();
+        let proof_encoded = self.encode();
+
+        ethabi::encode(&[
+            Token::Bytes(data_encoded.into()),
+            Token::Bytes(proof_encoded.into()),
+        ])
+        .into()
+    }
+}
+
+impl traits::CommandBatch for CommandBatch {
+    fn new(block_height: u64, messages: Vec<Message>, destination_chain_id: Uint256) -> Self {
+        let message_ids = messages.iter().map(|msg| msg.id.clone()).collect();
+
+        let data = Data::new(destination_chain_id, messages);
+        let encoded_data = data.encode();
+
+        let id = batch_id(block_height, &encoded_data);
+        let msg_to_sign = msg_to_sign(&encoded_data);
+
+        Self {
+            id,
+            message_ids,
+            data,
+            msg_to_sign,
+            multisig_session_id: None,
+        }
+    }
+}
+
+fn command_id(message_id: String) -> [u8; 32] {
+    // TODO: we might need to change the command id format to match the one in core for migration purposes
+
+    Keccak256::digest(message_id.as_bytes()).into()
+}
+
+// TODO: This will make it incompatible with current version of destination chain gateways,
+// they rely on tx hash and event index as well, but just to emit events
+// https://github.com/axelarnetwork/axelar-cgp-solidity/blob/main/contracts/AxelarGateway.sol#L466
+fn command_params(
+    source_chain: String,
+    source_address: String,
+    destination_address: ethereum_types::Address,
+    payload_hash: [u8; 32],
+) -> HexBinary {
+    ethabi::encode(&[
+        Token::String(source_chain),
+        Token::String(source_address),
+        Token::Address(destination_address),
+        Token::FixedBytes(payload_hash.into()),
+    ])
+    .into()
+}
+
+fn batch_id(block_height: u64, data: &HexBinary) -> HexBinary {
+    let mut id_hasher = Keccak256::new();
+
+    id_hasher.update(block_height.to_be_bytes());
+    id_hasher.update(data.as_slice());
+
+    id_hasher.finalize().as_slice().into()
+}
+
+fn msg_to_sign(data: &HexBinary) -> HexBinary {
+    let msg = Keccak256::digest(data.as_slice());
+
+    let unsigned = [
+        "\x19Ethereum Signed Message:\n32".as_bytes(), // Keccek256 hash length = 32
```

Comment:
Why do we add this prefix? Probably worth adding a comment.

Reply:
its the standard for EVM signed data: https://eips.ethereum.org/EIPS/eip-191. I just added the comment